dali/pipeline/data/tensor_vector.cc

```diff
@@ -60,20 +65,147 @@ TensorVector<Backend>::TensorVector(TensorVector<Backend> &&other) noexcept {
+  // TODO(klecki): more consistency checks, contiguous -> non-contiguous removes shares_data from
+  // samples
+  // Bounds check
+  assert(sample_idx >= 0 && sample_idx < static_cast<int>(curr_tensors_size_));
+  DALI_ENFORCE(type() == owner.type(),
+               make_string("Sample must have the same type as a target batch, current: ", type(),
+                           " new: ", owner.type(), " for ", sample_idx, "."));
+  DALI_ENFORCE(
+      sample_dim() == owner.shape().sample_dim(),
+      make_string("Sample must have the same dimensionality as a target batch, current: ",
+                  sample_dim(), " new: ", owner.shape().sample_dim(), " for ", sample_idx, "."));
+  DALI_ENFORCE(this->order() == owner.order(), "Sample must have the same order as a target batch");
+  DALI_ENFORCE(GetLayout() == owner.GetLayout(),
+               make_string("Sample must have the same layout as a target batch current: ",
+                           GetLayout(), " new: ", owner.GetLayout(), " for ", sample_idx, "."));
+  DALI_ENFORCE(
+      is_pinned() == owner.is_pinned(),
+      make_string("Sample must have the same pinned status as target batch, current: ", is_pinned(),
+                  " new: ", owner.is_pinned(), " for ", sample_idx, "."));
+  SetContiguous(false);
+  // Setting a new share overwrites the previous one - so we can safely assume that even if
+  // we had a sample sharing into TL, it will be overwritten
+  tensors_[sample_idx]->ShareData(owner);
+  tl_->Reset();
+}
+
+template <typename Backend>
+void TensorVector<Backend>::UnsafeSetSample(int sample_idx, const shared_ptr<void> &ptr,
+                                            size_t bytes, bool pinned, const TensorShape<> &shape,
+                                            DALIDataType type, AccessOrder order,
+                                            const TensorLayout &layout) {
+  assert(sample_idx >= 0 && sample_idx < static_cast<int>(curr_tensors_size_));
+  DALI_ENFORCE(this->type() == type,
+               make_string("Sample must have the same type as a target batch, current: ",
+                           this->type(), " new: ", type, " for ", sample_idx, "."));
+  DALI_ENFORCE(sample_dim() == shape.sample_dim(),
+               make_string("Sample must have the same dimensionality as a target batch, current: ",
+                           sample_dim(), " new: ", shape.sample_dim(), " for ", sample_idx, "."));
+  DALI_ENFORCE(this->order() == order, "Sample must have the same order as a target batch");
+  DALI_ENFORCE(GetLayout() == layout,
+               make_string("Sample must have the same layout as a target batch current: ",
+                           GetLayout(), " new: ", layout, " for ", sample_idx, "."));
+  DALI_ENFORCE(is_pinned() == pinned,
+               make_string("Sample must have the same pinned status as target batch, current: ",
+                           is_pinned(), " new: ", pinned, " for ", sample_idx, "."));
```

Comment:
Would it make sense to place it in a separate method and reuse across all Unsafe(...) methods? Kind of like with `SetupLikeImpl`?